### DISCUSSION TITLE
Switch IRavenQueryable<T> to IRavenQueryable<out T>

### DISCUSSION THREAD

**Discussion by stevehansen** [ravendb/ravendb#15604] [Ideas]
```IRavenQueryable<T>``` implements ```IOrderedQueryable<T>, IEnumerable<T>, IQueryable<T>``` all three having out T
We got bit by this because our T was an interface so the casts that worked for IQueryable don't work for IRavenQueryable

**Comment by ayende** [comment]
Sounds reasonable, but what is the impact on backward compact?

**Reply by stevehansen** [reply]
https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/covariance-contravariance/variance-in-generic-interfaces we do have it for a long time in .NET, I don't expect it to break any existing code, the only difference would be that we can now pass IRavenQueryable<X> to any base classes of X or implemented interfaces

public void Handle(IRavenQueryable<object> source) would now be able to accept any generic queryable

https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/covariance-contravariance/using-variance-in-interfaces-for-generic-collections#converting-generic-collections

This would probably describe our use case, an alternative would have been to do public void Handle<TEntity>(IRavenQueryable<TEntity> source) where TEntity : ISomeInterface
but in our case we didn't have the generic part, we got an IQueryable<TInterface> as argument and couldn't cast it to IRavenQueryable<TInterface> because it was actually an IRavenQueryable<TEntity>

**Reply by ayende** [reply]
The problem is that you may have a user that is using `IRavenQueryable` and whatever the `out T` would surprise them here.
